### PR TITLE
Feature/active record wallet adapter

### PR DESCRIPTION
**Pull request by Yamaguchi**
This PR is solution for a part of #26

- Implement ActiveRecordWalletAdapter as a class inherited from AbstractWalletAdapter, and
- Implement Wallet, Key, Utxo classes as an ActiveRecord
- Implement generators for the migration database tables.

Tasks listed below are out of scope.

- Implement state transition
    - Utxo should be unspendable when the transaction will use it as an input.
    - Utxo should change its state to "broadcasted" when the transaction will be broadcasted.
    - Utxo should change its state to "finalized" when the transaction will be finalized.
    - Utxo should be deleted when the transaction that uses it as an input will be finalized.
- Implement other utilities.
    - rake task to synchronize tapyrus blockchain and import Utxos from external transactions.

### CODE REVIEW

File: lib/generators/glueby/contract/templates/utxo_table.rb.erb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+      t.integer    :index
+      t.bigint     :value
+      t.string     :script_pubkey
+      t.boolean    :spent
+      t.integer    :status
+      t.belongs_to :key, null: true
+      t.timestamps
+    end
+
+    add_index :utxos, [:txid, :index]
```
Comment (azuchi) [inline lib/generators/glueby/contract/templates/utxo_table.rb.erb:14]: Isn't it better to have a unique constraint on txid and index?

File: lib/generators/glueby/contract/templates/key_table.rb.erb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    create_table :keys<%= table_options %> do |t|
+      t.string     :private_key
+      t.string     :public_key
+      t.string     :script_pubkey
+      t.integer    :purpose
+      t.belongs_to :wallet
+      t.timestamps
+    end
+
+    add_index :keys, [:script_pubkey]
```
Comment (azuchi) [inline lib/generators/glueby/contract/templates/key_table.rb.erb:12]: Isn't it better to have a unique constraint on private_key?

File: lib/glueby/internal/wallet/active_record/wallet.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+          def key_for_input(input)
+            out_point = input.out_point
+            utxo = Glueby::Internal::Wallet::AR::Utxo.find_by(txid: out_point.txid, index: out_point.index)
+            return unless utxo
+            Key.find_by(script_pubkey: utxo.script_pubkey)
+          end
+
+          def sign_tx_for_p2pkh(tx, index, key)
+            sighash = tx.sighash_for_input(index, key.to_p2pkh)
+            sig = key.sign(sighash) + [Tapyrus::SIGHASH_TYPE[:all]].pack('C')
```
[the remark this answers is not in the document]
Comment (Yamaguchi) [reply lib/glueby/internal/wallet/active_record/wallet.rb:37]: Fix at 2c4f51a71b6ab980d45ac93bab620f7da6ae2c82

File: lib/glueby/internal/wallet/active_record/utxo.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        class Utxo < ::ActiveRecord::Base
+          belongs_to :key
+
+          enum status: { init: 0, broadcasted: 1, finalized: 2 }
+
+          def color_id
+            script = Tapyrus::Script.parse_from_payload(script_pubkey.htb)
+            if script.cp2pkh? || script.cp2sh?
+              Tapyrus::Color::ColorIdentifier.parse_from_payload(script.chunks[0].pushed_data).to_hex
+            end
```
[the remark this answers is not in the document]
Comment (Yamaguchi) [reply lib/glueby/internal/wallet/active_record/utxo.rb]: Updated ccf35ed2abeb463c69d33ccd322e3178dfa8e1d8.

[the remark this answers is not in the document]
Comment (Yamaguchi) [reply lib/glueby/internal/wallet/active_record/utxo.rb]: But it breaks rspec until tapyrus 0.2.5 is released.

File: lib/generators/glueby/contract/templates/utxo_table.rb.erb (hunk 1 above)
Comment (Yamaguchi) [reply lib/generators/glueby/contract/templates/utxo_table.rb.erb]: Added at 8ca322ed89ab1bb551e953f8b6d52cfe68490e4f

File: lib/generators/glueby/contract/templates/key_table.rb.erb (hunk 2 above)
Comment (Yamaguchi) [reply lib/generators/glueby/contract/templates/key_table.rb.erb]: Added at 8ca322ed89ab1bb551e953f8b6d52cfe68490e4f

File: lib/generators/glueby/contract/templates/utxo_table.rb.erb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+class CreateUtxo < ActiveRecord::Migration<%= migration_version %>
+  def change
+    create_table :utxos<%= table_options %> do |t|
+      t.string     :txid
+      t.integer    :index
+      t.bigint     :value
+      t.string     :script_pubkey
+      t.boolean    :spent
```
[the remark this answers is not in the document]
Comment (Yamaguchi) [reply lib/generators/glueby/contract/templates/utxo_table.rb.erb]: Remove spent column(318bd0e).

**Pull request merged**